packages/core/src/checkout-buttons/strategies/paypal-commerce/paypal-commerce-button-strategy.ts

```diff
@@ -89,21 +114,202 @@ export default class PaypalCommerceButtonStrategy implements CheckoutButtonStrat
             const cartRequestBody = buyNowInitializeOptions.getBuyNowCartRequestBody();
-
             if (!cartRequestBody) {
                 throw new MissingDataError(MissingDataErrorType.MissingCart);
             }
 
             try {
                 const { body: cart } = await this._cartRequestSender.createBuyNowCart(cartRequestBody);
-
                 this._buyNowCartId = cart.id;
             } catch (error) {
                 throw new BuyNowCartCreationError();
             }
         }
     }
 
+
+    private async _onCancel() {
+        const state = this._store.getState();
+        const billingAddress = state.billingAddress.getBillingAddressOrThrow();
+
+        if (billingAddress) {
+            const resetAddress = this._resetAddress(billingAddress);
+            await this._store.dispatch(this._billingAddressActionCreator.updateAddress(resetAddress));
+            await this._store.dispatch(this._consignmentActionCreator.updateAddress(resetAddress));
+
+            await this._updateOrder();
+        }
+    }
+
+    private _resetAddress(address: BillingAddressRequestBody) {
+        const { firstName, lastName, address1, email } = address;
+
+        return {
+            ...address,
+            firstName: firstName !== 'Fake' ? firstName : '',
+            lastName: lastName !== 'Fake' ? lastName : '',
+            address1: address1 !== 'Fake street' ? address1 : '',
+            email: email !== 'fake@fake.fake' ? email : '',
+        }
+    }
+
+    private async _onHostedCheckoutApprove(
+        data: ApproveCallbackPayload,
+        actions: ApproveCallbackActions,
+        methodId: string,
+    ): Promise<boolean> {
+        const state = this._store.getState();
+        const cart = state.cart.getCartOrThrow();
+        const orderDetails = await actions.order.get();
+
+        if (cart.lineItems.physicalItems.length > 0) {
+            const address = this._getValidAddress(
+                orderDetails.payer.name,
+                orderDetails.payer.email_address,
+                orderDetails.purchase_units[0].shipping.address
+            );
+
+            await this._store.dispatch(this._billingAddressActionCreator.updateAddress(address));
+            await this._store.dispatch(this._consignmentActionCreator.updateAddress(address));
+            await this._updateOrder();
+        } else {
+            const address = this._getValidAddress(
+                orderDetails.payer.name,
+                orderDetails.payer.email_address,
+                orderDetails.payer.address
+            );
+
+            await this._store.dispatch(this._billingAddressActionCreator.updateAddress(address));
+        }
+
+        await this._store.dispatch(this._orderActionCreator.submitOrder({}, { params: { methodId } }));
+        await this._submitPayment(methodId, data.orderID);
+        window.location.assign('/checkout/order-confirmation');
```

Comment:
We should avoid using window.location.assign in checkout-sdk.

This code will cause an issue on BC side because the order will not be finished for us.

The redirect logic is located in onComplete method where **callback** is called + some extra logic with secondary submitting payment. The **callback** is a function that passed from BCapp stencil template and makes this redirect when called.

Reply:
fixed